Docs: Update README to reflect pre-preview status and functionality (#16)
xmtp/xmtp-react-native: README.md

# xmtp-react-native-sdk

![Lint](https://github.com/xmtp/xmtp-android/actions/workflows/lint.yml/badge.svg) ![Status](https://img.shields.io/badge/Project_Status-Pre--Preview-red)

<!--eng: Correctly stated below? Necessary to say which native XMTP SDKs, or no?-->

> This repo provides wrappers that help you build a React Native app with XMTP. The wrappers serve as a bridge between your React Native app and native XMTP SDKs.

Reviewer: ```suggestion
This repo provides a package you can use to build with XMTP in a React Native or Expo app.
```

I removed the mentions of the wrappers since that's just an implementation detail that devs shouldn't need to worry about.

> **Important:**  
> This SDK is in **Pre-Preview** status and ready for you to start experimenting with.
>
> However, we do **not** recommend using Pre-Preview software in production apps. Software in this status will change as we iterate on features and feedback.

> Specifically, this SDK is missing this functionality:
> 
> <!--eng: Fill in this list below, if applicable. Which features are missing that we might want to call out? in the second bullet, our pattern has been to call out the lack of parity between the SDK and xmtp-js. For this React Native SDK, is the comparison to xmtp-js still useful? Or does the comparison need to be with the xmtp-ios and xmtp-android SDKs? Or maybe parity with another SDK is not something we need to call out?-->
> 
> - The SDK connects to the `dev` environment only. Support for connecting to `production` and `local` environments is pending.
> 
> - The `apiUrl`, `keyStoreType`, `codecs`, `maxContentSize` and `appVersion` parameters present in the XMTP client SDK for JavaScript (xmtp-js) are not yet supported.
> 
> - Something else not supported
> 
> To keep up with the latest SDK developments, see the [Issues tab](https://github.com/xmtp/xmtp-react-native/issues) in this repo.
> 
> See an issue you'd like to work on? XMTP welcomes contributions! If you contribute a PR that gets merged into this repo, you'll be eligible to [claim this XMTP contributor POAP](https://www.gitpoap.io/gp/1042).

Reviewer: ```suggestion
Specifically, this SDK is missing this functionality:

- Basically everything (lol feel free to remove this jha, the comparison we want is between xmtp-js with the eventual goal to be a complete match in the APIs)
- The SDK connects to the `dev` environment only. Support for connecting to `production` and `local` environments is pending.
- The `apiUrl`, `keyStoreType`, `codecs`, `maxContentSize` and `appVersion` parameters present in the XMTP client SDK for JavaScript (xmtp-js) are not yet supported.

Here's what actually works right now:

- Creating a `Client` from an ethers `Signer`
- Listing conversations
- Listing message (unpaginated for now) from a conversation
- Starting a conversation

We have a very incomplete list of TODOs here: https://github.com/xmtp/xmtp-react-native/issues. If there's anything you need that is missing from that list, please open an issue, or better yet, open a PR and we'll get it reviewed and merged as soon as possible!

See an issue you'd like to work on? XMTP welcomes contributions! If you contribute a PR that gets merged into this repo, you'll be eligible to [claim this XMTP contributor POAP](https://www.gitpoap.io/gp/1042).
```

To learn more about XMTP and get answers to frequently asked questions, see [FAQ about XMTP](https://xmtp.org/docs/dev-concepts/faq).

![x-red-sm](https://user-images.githubusercontent.com/510695/163488403-1fb37e86-c673-4b48-954e-8460ae4d4b05.png)

## Example apps

This repo provides the following example apps:

- [Android example](example/android)

- [iOS example](example/ios)

<!--eng: is this command below correct? Do I just run it in the SDK's root dir?-->

To use an example app, run:

```bash
npm install --force

> ```

Reviewer: ```suggestion
To use an example app, run:

```bash
cd example
npm install --force
npm run [ios or android]
```
```

## API documentation

> <!--eng: Relevant for this repo? Both links are 404s right now. Will these doc links eventually provide access to SDK docs generated from code annotations added to this repo?-->

Reviewer: I'm not sure what the story is for generated docs in a React Native app. The _hope_ is that the docs for the `xmtp-js` SDK should "Just work" for the React native sdk.

Author: Ah, this is great to know, @nakajima! I'm going to remove this API documentation section for now as the links don't work and I don't think we're ready to point to the xmtp-js SDK reference docs just yet. This is just fine for Pre-Preview status. I think once we move to Dev Preview we can talk about getting our reference docs plan in place for GA.

- [Documentation for the main branch](https://github.com/expo/expo/blob/main/docs/pages/versions/unversioned/sdk/xmtp-react-native-sdk.md)

- [Documentation for the latest stable release](https://docs.expo.dev/versions/latest/sdk/xmtp-react-native-sdk/)

> <!--eng: I basically created two paths for install - one for a managed Expo project and one for a bare RN project. The dev takes one route or the other depending on their setup, is that correct?-->

Reviewer: > The dev takes one route or the other depending on their setup, is that correct?

That's correct. Everything under the "NOTES" section from before was generated by the package we used to make this project so I'm not super familiar with the specifics.

## Install in a managed Expo project

> <!--eng: Is this correct? There is no API documentation for this right now - so is install for an expo project not possible yet? Or do we have a command we want to share here?-->

Reviewer: I _think_ managed expo projects are ones where the iOS/Android parts are hidden away, which doesn't work for this package because we're wrapping our native SDKs. I think people using a managed expo project will need to run `npx expo prebuild`.

For a [managed](https://docs.expo.dev/versions/latest/introduction/managed-vs-bare/) Expo project, follow the installation instructions in the [API documentation](#api-documentation) for the latest stable release. If you follow the link and there is no documentation available, then this library is not yet usable within managed projects. It is likely to be included in an upcoming Expo SDK release.

## Install and configure in a bare React Native project

> <!--eng: What does this mean - why do I need to install the Expo package if this is a bare RN project? Is there a command we want to share here?-->

Reviewer: Expo sets up some additional stuff around wrapping native modules that RN doesn't do on its own. I think the link to the docs (https://docs.expo.dev/bare/installing-expo-modules/) should be a good place to point people.

For bare React Native projects, you must ensure that you have [installed and configured the `expo` package](https://docs.expo.dev/bare/installing-expo-modules/) before continuing.

### Add the package to your npm dependencies

```bash
npm install xmtp-react-native-sdk
```

### Configure for iOS

<!--eng: For this command and the one below - what are we configuring for iOS and for Android? When does this configuration happen? Do I run these commands in the root dir?-->

After installing the npm package, run:

```bash
npx pod-install

> ```

Reviewer: These are great questions and I do not know the answers to them, the instructions came from the expo module skeleton so I'd say probably worth keeping them around?

### Configure for Android

After installing the npm package, run:

<!--eng: Just guessing at this based on talking with ChatGPT LOL. Says that starting from React Native 0.60, this command may not be necessary.-->

```bash
react-native link xmtp-react-native-sdk

> ```

Reviewer: I don't think we need this, I think expo handles linking for us?

> ## Usage overview

Reviewer: I think while stuff is still in this super early phase, since a lot of the APIs from xmtp-js are going to be missing, I'd say maybe we just point people to the [example app](https://github.com/xmtp/xmtp-react-native/tree/main/example/src) for usage. Once we've got stuff more fleshed out, we can add this section (which should basically just mirror the usage from xmtp/xmtp-js).

Currently, these are the features supported by this SDK:

- Create a `Client` from an ethers `Signer`
- List conversations
- List messages (unpaginated for now) in a conversation
- Start a conversation

Here's a [list of features and issues](https://github.com/xmtp/xmtp-react-native/issues) we're still working on. 

Is there a feature you need that's not on the list? Please [open an issue](https://github.com/xmtp/xmtp-react-native/issues).

Or better yet, open a PR and we'll get it reviewed and merged as soon as possible. If you contribute a PR that gets merged into this repo, you'll be eligible to [claim this XMTP contributor POAP](https://www.gitpoap.io/gp/1042)!

To learn more about XMTP and get answers to frequently asked questions, see [FAQ about XMTP](https://xmtp.org/docs/dev-concepts/faq).

![x-red-sm](https://user-images.githubusercontent.com/510695/163488403-1fb37e86-c673-4b48-954e-8460ae4d4b05.png)

## Example apps

This repo provides the following example apps that you can use to experiment and explore implementation details:

- [Android example](example/android)

> - [iOS example](example/ios)

Reviewer 2: I think the example app is just in example. There shouldn't be any need for them to go into the respective ios or android folders.

To use an example app, run:

```bash
cd example
npm install --force
npm run [ios or android]
```

## Install in a managed Expo project

> <!--eng: When I run this at the root of the project, I get this prompt: The android and ios projects are malformed, would you like to clear the project files and reinitialize them? › (Y/n) I imagine this is because I'm not running it in a managed Expo project. Just wanted to run it to see what happened. =)-->

Reviewer: Ya this is a command you run from inside an actual app, not the library we've got here.

```bash
npx expo prebuild
```

## Install in bare React Native project

For bare React Native projects, [install and configure the `expo` package](https://docs.expo.dev/bare/installing-expo-modules/) before continuing.

### Add the package to your npm dependencies

> <!--eng: Should this command use xmtp-react-native or xmtp-react-native-sdk? When I run this command, I get npm ERR! 404 Not Found - GET https://registry.npmjs.org/xmtp-react-native - Not found. This is because we haven't submitted to npm yet, correct? Will this command work for pre-preview?-->

Reviewer: Once https://github.com/xmtp/xmtp-react-native/pull/12 lands, it should be npm installable. It's not yet though!

```bash
npm install xmtp-react-native-sdk
```

### Configure for iOS

After installing the npm package, run:

```bash
npx pod-install
```

### Configure for Android

> <!--eng: Seems weird to have a section to say that there's nothing to do, but thought it might be helpful to explictly call this out. If we leave it out, folx might think the Android info is missing. WDYT?-->

Reviewer: I think once we're published to npm and can E2E test the installation flow, we'll have more to say about platform specific setup. It's sort of an option question about how this stuff works right now to me.

For example, for a given blockchain account address, its XMTP identity on `dev` network is completely distinct from its XMTP identity on the `production` network, as are the messages associated with these identities. In addition, XMTP identities and messages created on the `dev` network can't be accessed from or moved to the `production` network, and vice versa.

**Important:** When you create a client, it connects to the XMTP `dev` environment by default. You can use the `env` parameter to set your client's network environment.

The `env` parameter accepts one of three valid values: `dev`, `production`, or `local`. Here are some best practices for when to use each environment:

- `dev`: Use to have a client communicate with the `dev` network. As a best practice, set `env` to `dev` while developing and testing your app. Follow this best practice to isolate test messages to `dev` inboxes.

- `production`: Use to have a client communicate with the `production` network. As a best practice, set `env` to `production` when your app is serving real users. Follow this best practice to isolate messages between real-world users to `production` inboxes.

> - `local`: Use to have a client communicate with an XMTP node you are running locally. For example, an XMTP node developer can set `env` to `local` to generate client traffic to test a node running locally.

Reviewer: FWIW, none of this currently works in this project (tracked in https://github.com/xmtp/xmtp-react-native/issues/4).
